### PR TITLE
IGAPP-383: Additional field for feedback contact mail

### PR DESCRIPTION
**Pull request by Cmd190**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

Feedback-View in the webapp:

image

Feedback in CMS:

image

### CODE REVIEW

**Comment by Cmd190** [comment]
I updated the tests but for some reason flow declares an error for on file:

Error ----------------------------------------------- src/modules/feedback/components/__tests__/FeedbackBox.spec.js:25:7

Cannot create `FeedbackBox` element because property `onContactMailChanged` is missing in props [1] but exists in
`PropsType` [2].

   src/modules/feedback/components/__tests__/FeedbackBox.spec.js:25:7
             v-----------
   25|       <FeedbackBox
   26|         isPositiveRatingSelected={false}
   27|         comment=''
   29|         selectedFeedbackOption={feedbackOptions[0]}
   30|         onCommentChanged={onCommentChanged}
   31|         onFeedbackOptionChanged={onFeedbackOptionChanged}
   32|         onSubmit={onSubmit}
   33|         sendingStatus='SUCCESS'
   34|         t={t}
   35|         closeFeedbackModal={() => {}}
   36|         theme={theme} />
               ---------------^ [1]

Can someone help me with this issue?

**Comment by klinzo** [comment]
> I updated the tests but for some reason flow declares an error for on file:
> 
> Error ----------------------------------------------- src/modules/feedback/components/**tests**/FeedbackBox.spec.js:25:7
> 
> Cannot create `FeedbackBox` element because property `onContactMailChanged` is missing in props [1] but exists in
> `PropsType` [2].
> 
> src/modules/feedback/components/**tests**/FeedbackBox.spec.js:25:7
> v-----------
> 25| <FeedbackBox
> 26| isPositiveRatingSelected={false}
> 27| comment=''
> 29| selectedFeedbackOption={feedbackOptions[0]}
> 30| onCommentChanged={onCommentChanged}
> 31| onFeedbackOptionChanged={onFeedbackOptionChanged}
> 32| onSubmit={onSubmit}
> 33| sendingStatus='SUCCESS'
> 34| t={t}
> 35| closeFeedbackModal={() => {}}
> 36| theme={theme} />
> ---------------^ [1]
> 
> Can someone help me with this issue?

You need to pass the `onContactMailChanged` prop to fix the flow error.

File: translations/translations.json [hunk 1]
```diff
@@ -1060,7 +1060,8 @@
       "thanksMessage": "Vielen Dank für Ihr Feedback. Damit helfen Sie uns, unsere Inhalte und die {{appName}}-App zu verbessern. Wir würden uns über weiteres Feedback freuen.",
       "close": "Schließen",
       "negativeRating": "Diese Seite ist nicht hilfreich",
-      "positiveRating": "Diese Seite ist hilfreich"
+      "positiveRating": "Diese Seite ist hilfreich",
+      "contactMailAdress": "Email für eventuelle Rückfragen (optional)"
```
Comment (klinzo) [inline translations/translations.json:1064]:
```suggestion
      "contactMailAdress": "E-Mail für eventuelle Rückfragen (optional)"
```
Maybe we should translate `optional` separately and combine it :thinking:

File: translations/translations.json [hunk 2]
```diff
@@ -1114,7 +1115,8 @@
       "thanksMessage": "Thank you for your feedback. This will help us to improve our content and the {{appName}} app. We would be happy to receive further feedback.",
       "close": "Close",
       "negativeRating": "This site is not useful",
-      "positiveRating": "This site is useful"
+      "positiveRating": "This site is useful",
+      "contactMailAdress": "Email for possible queries (non-mandatory)"
```
Comment (klinzo) [inline translations/translations.json:1119]:
```suggestion
```
Just leave it out and we'll let it translate professionally :)

File: web/src/modules/feedback/components/FeedbackBox.js [hunk 3]
```diff
@@ -75,8 +81,12 @@ export class FeedbackBox extends React.PureComponent<PropsType> {
           commentMessage={isPositiveRatingSelected ? t('positiveComment') : t('negativeComment')}
           onCommentChanged={onCommentChanged}
           required={!isPositiveRatingSelected} />
+        <Description>{t('contactMailAdress')}</Description>
+        <TextInput
+          onChange={onContactMailChanged}
+        />
         {sendingStatus === 'ERROR' && <Description>{t('failedSendingFeedback')}</Description>}
-        <TextButton
+         <TextButton sfd
```
Comment (klinzo) [inline web/src/modules/feedback/components/FeedbackBox.js:89]: Why this change?

File: web/src/modules/feedback/components/FeedbackBox.js [hunk 4]
```diff
@@ -75,8 +81,12 @@ export class FeedbackBox extends React.PureComponent<PropsType> {
           commentMessage={isPositiveRatingSelected ? t('positiveComment') : t('negativeComment')}
           onCommentChanged={onCommentChanged}
           required={!isPositiveRatingSelected} />
+        <Description>{t('contactMailAdress')}</Description>
+        <TextInput
+          onChange={onContactMailChanged}
+        />
```
Comment (klinzo) [inline web/src/modules/feedback/components/FeedbackBox.js:87]: Can't we reuse `FeedbackComment` for this?

File: web/src/modules/layout/components/__tests__/__snapshots__/Header.spec.js.snap [hunk 5]
```diff
@@ -4,16 +4,9 @@ exports[`Header should match snapshot with largeViewport 1`] = `
 <Headroom
   height={180}
   onStickyTopChanged={[Function]}
-  parent={
```
Comment (klinzo) [inline web/src/modules/layout/components/__tests__/__snapshots__/Header.spec.js.snap:7]: Any idea why these snapshots have been changed?

**Review by klinzo — CHANGES_REQUESTED** [review]
:taco: :mailbox_with_mail: :tada:

File: web/src/modules/feedback/components/FeedbackBox.js (hunk 3 above)
Comment (Cmd190) [reply web/src/modules/feedback/components/FeedbackBox.js:89]: Well, I would call that a bug 😂

File: translations/translations.json (hunk 2 above)
Comment (Cmd190) [reply translations/translations.json:1119]: ok

File: web/src/modules/layout/components/__tests__/__snapshots__/Header.spec.js.snap (hunk 5 above)
Comment (Cmd190) [reply web/src/modules/layout/components/__tests__/__snapshots__/Header.spec.js.snap:7]: nope... I just ran test update command

File: web/src/modules/feedback/components/FeedbackBox.js (hunk 4 above)
Comment (Cmd190) [reply web/src/modules/feedback/components/FeedbackBox.js:87]: Imo comment and contactMail should be seperated because they just are two different entries.

File: translations/translations.json [hunk 6]
```diff
@@ -1060,7 +1060,9 @@
       "thanksMessage": "Vielen Dank für Ihr Feedback. Damit helfen Sie uns, unsere Inhalte und die {{appName}}-App zu verbessern. Wir würden uns über weiteres Feedback freuen.",
       "close": "Schließen",
       "negativeRating": "Diese Seite ist nicht hilfreich",
-      "positiveRating": "Diese Seite ist hilfreich"
+      "positiveRating": "Diese Seite ist hilfreich",
+      "contactMailAdress": "E-Mail für eventuelle Rückfragen",
```
Comment (maxammann) [inline translations/translations.json:1064]:
```suggestion
      "contactMailAdress": "E-Mail für Rückfragen",
```

Translations are easier for that one.

Comment (maxammann) [inline translations/translations.json:1064]:
```suggestion
      "contactMailAddress": "E-Mail für eventuelle Rückfragen",
```

File: web/src/modules/feedback/components/FeedbackBoxContainer.js [hunk 7]
```diff
@@ -237,9 +240,10 @@ export class FeedbackBoxContainer extends React.Component<PropsType, StateType>
 
   handleSubmit = async () => {
     const { onSubmit } = this.props
-    const { selectedFeedbackOption, comment } = this.state
+    const { selectedFeedbackOption, comment, contactMail } = this.state
     try {
-      await this.postFeedbackData(this.getFeedbackData(selectedFeedbackOption, comment))
+      await this.postFeedbackData(
+        this.getFeedbackData(selectedFeedbackOption, `${comment}    contactMail: ${contactMail}`))
```
Comment (maxammann) [inline web/src/modules/feedback/components/FeedbackBoxContainer.js:246]:
```suggestion
        this.getFeedbackData(selectedFeedbackOption, `${comment}\m\nKontakt Adresse: ${contactMail}`))
```

I suggest to make this string german as all people who read comments are German

Comment (maxammann) [inline web/src/modules/feedback/components/FeedbackBoxContainer.js:246]:
```suggestion
        this.getFeedbackData(selectedFeedbackOption, `${comment}    contactMail: ${contactMail || 'Keine'}`))
```

Comment (maxammann) [reply web/src/modules/feedback/components/FeedbackBoxContainer.js:246]: If there is no contactMail then add a meaningful word instead of an empty string.

**Review by maxammann — CHANGES_REQUESTED** [review]

File: web/src/modules/feedback/components/FeedbackBoxContainer.js (hunk 7 above)
Comment (Cmd190) [reply web/src/modules/feedback/components/FeedbackBoxContainer.js:246]: Yes, you're right

File: web/src/modules/feedback/components/FeedbackBox.js (hunk 4 above)
Comment (Cmd190) [reply web/src/modules/feedback/components/FeedbackBox.js]: @klinzo so what should we do?

Comment (Taggotty) [reply web/src/modules/feedback/components/FeedbackBox.js]: I think you could refactor `FeedbackComment` to sth like `FeedbackText` or `FeedbackEntry` (any name that fits)

File: web/src/modules/feedback/components/FeedbackBoxContainer.js [hunk 8]
```diff
@@ -237,9 +240,10 @@ export class FeedbackBoxContainer extends React.Component<PropsType, StateType>
 
   handleSubmit = async () => {
     const { onSubmit } = this.props
-    const { selectedFeedbackOption, comment } = this.state
+    const { selectedFeedbackOption, comment, contactMail } = this.state
     try {
-      await this.postFeedbackData(this.getFeedbackData(selectedFeedbackOption, comment))
+      await this.postFeedbackData(
+        this.getFeedbackData(selectedFeedbackOption, `${comment}\m\nKontakt Adresse: ${contactMail}`))
```
Comment (klinzo) [inline web/src/modules/feedback/components/FeedbackBoxContainer.js:246]:
```suggestion
        this.getFeedbackData(selectedFeedbackOption, `${comment} Kontaktadresse: ${contactMail}`))
```
otherwise it looks quite weird as linebreaks are not working.

**Review by klinzo — APPROVED** [review]
Tested on firefox, looking good :+1: You could look into whether it makes sense to generalize `FeedbackComment` as @Taggotty suggested, but I fear it does not make sense as one is a TextInput whereas the other is a TextArea. Your choice I'd say @Cmd190

**Comment by Cmd190** [comment]
> Tested on firefox, looking good 👍 You could look into whether it makes sense to generalize `FeedbackComment` as @Taggotty suggested, but I fear it does not make sense as one is a TextInput whereas the other is a TextArea. Your choice I'd say @Cmd190



> Tested on firefox, looking good 👍 You could look into whether it makes sense to generalize `FeedbackComment` as @Taggotty suggested, but I fear it does not make sense as one is a TextInput whereas the other is a TextArea. Your choice I'd say @Cmd190

Yes, that's the point. I'll leave it as it is.

File: package-lock.json [hunk 9]
```diff
@@ -0,0 +1,2190 @@
+{
```
Comment (klinzo) [inline package-lock.json:1]: This should not be included as we are using yarn and not npm to manage our packages. Please remove again :)

**Review by klinzo — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 9 above)
Comment (Cmd190) [reply package-lock.json:1]: I know. There is another file too which should not be included into the commit. I'll take care of this

File: web/src/modules/feedback/components/FeedbackBoxContainer.js (hunk 8 above)
Comment (maxammann) [reply web/src/modules/feedback/components/FeedbackBoxContainer.js]: @Cmd190 and I discussed this in mattermost. For now we will go with the white-space.

Comment (klinzo) [reply web/src/modules/feedback/components/FeedbackBoxContainer.js]: @maxammann so the way it is currently?

File: web/src/modules/feedback/components/FeedbackBoxContainer.js [hunk 10]
```diff
@@ -234,9 +237,10 @@ export class FeedbackBoxContainer extends React.Component<PropsType, StateType>
 
   submitFeedback = async () => {
     const { onSubmit } = this.props
-    const { selectedFeedbackOption, comment } = this.state
+    const { selectedFeedbackOption, comment, contactMail } = this.state
     try {
-      await this.postFeedbackData(this.getFeedbackData(selectedFeedbackOption, comment))
+      await this.postFeedbackData(
+        this.getFeedbackData(selectedFeedbackOption, `${comment}    Kontaktadresse: ${contactMail || 'Keine Angabe'}`))
```
Comment (maxammann) [inline web/src/modules/feedback/components/FeedbackBoxContainer.js:243]: Make sure that this line is equal to that of the react-native app

**Review by maxammann — APPROVED** [review]

**Review by klinzo — APPROVED** [review]

**Review by Znippet — APPROVED** [review]

**Review by klinzo — APPROVED** [review]
